### PR TITLE
feat: allow passing functions in `setValidationEnabled`

### PR DESCRIPTION
**Pull request by imranbarbhuiya**

### CODE REVIEW

File: tests/lib/configs.test.ts [hunk 1]
```diff
@@ -11,6 +11,17 @@ describe('Validation enabled and disabled configurations', () => {
 	const setPredicate = s.set(s.number);
 	const tuplePredicate = s.tuple([s.string, s.number]);
 
+	const predicateAndValues = [
```
Comment (c43721) [inline tests/lib/configs.test.ts:14]: this looks like a minor side change, but I would suggest to do this in a separate PR to "reduce scope", but that's just me.

File: src/validators/BaseValidator.ts [hunk 2]
```diff
@@ -76,18 +77,18 @@ export abstract class BaseValidator<T> {
 	 * Sets if the validator should also run constraints or just do basic checks.
 	 * @param isValidationEnabled Whether this validator should be enabled or disabled. Set to `null` to go off of the global configuration.
 	 */
-	public setValidationEnabled(isValidationEnabled: boolean | null): this {
+	public setValidationEnabled(isValidationEnabled: boolean | (() => boolean) | null): this {
 		const clone = this.clone();
 		clone.isValidationEnabled = isValidationEnabled;
 		return clone;
```
Comment (c43721) [inline src/validators/BaseValidator.ts:83]: I would also add onto the doc comment that this accepts a function that returns a boolean.

File: tests/lib/configs.test.ts (hunk 1 above)
Comment (imranbarbhuiya) [reply tests/lib/configs.test.ts:14]: I'm using this array in my new test so I don't think it's out of scope.

**Review by favna — COMMENTED** [review]
I don't particularly mind the change but what is the use case of this exactly?

File: src/validators/BaseValidator.ts (hunk 2 above)
Comment (favna) [reply src/validators/BaseValidator.ts:83]: 🖕

Comment (c43721) [reply src/validators/BaseValidator.ts:83]: 👆this is the righ emoji

**Comment by imranbarbhuiya** [comment]
> I don't particularly mind the change but what is the use case of this exactly?

If you check my djs pr, I'm adding `setValidatorEnabled` every time before parsing. By allowing passing function, I can add it when defining the predicate and not every time before parsing.

**Comment by vladfrangu** [comment]
> By allowing passing function, I can add it when defining the predicate and not every time before parsing.

Not really true... it'll call the function when the setValidationEnabled is called. You probably want to set the internal field to the function instead and have it called in the getter

File: src/validators/BaseValidator.ts [hunk 3]
```diff
@@ -76,18 +77,18 @@ export abstract class BaseValidator<T> {
...
 	}
 
 	public getValidationEnabled() {
-		return this.isValidationEnabled;
+		return getValue(this.isValidationEnabled);
 	}
 
 	protected get shouldRunConstraints(): boolean {
-		return this.isValidationEnabled ?? getGlobalValidationEnabled();
+		return getValue(this.isValidationEnabled) ?? getGlobalValidationEnabled();
```
Comment (vladfrangu) [inline src/validators/BaseValidator.ts:91]:
```suggestion
		return this.isValidationEnabled === null ? getGlobalValidationEnabled() : getValue(this.isValidationEnabled) ?? getGlobalValidationEnabled();
```

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: src/validators/BaseValidator.ts (hunk 3 above)
Comment (imranbarbhuiya) [reply src/validators/BaseValidator.ts:91]: if `isValidationEnabled` is null, `getValue(this.isValidationEnabled)` will return null so it'll go to `getGlobalValidationEnabled()` so do we need an extra check?

Comment (vladfrangu) [reply src/validators/BaseValidator.ts:91]: oh right.. derpy moments

**Comment by imranbarbhuiya** [comment]
> 

Ig I'm doing the same

**Review by vladfrangu — DISMISSED** [review]

**Review by vladfrangu — APPROVED** [review]

**Review by favna — APPROVED** [review]
